Driver for the Wolfson WM8904 audio codec

VitekST opened the pull request:
This PR brings a driver for the Wolfson/Cirrus WM8904 audio codec, based on original work of @hakehuang (recently closed PR: https://github.com/zephyrproject-rtos/zephyr/pull/66609#event-13467395343). The driver was developed and tested on the `mimxrt595_evk` and `mimxrt685_evk` boards. Following improvements were made to the driver itself:
- decoupled from `hal_nxp`
- input routing (multiplexing)
- input mute and volume control
- output mute and volume control
- omission of dead code

Following modifications were needed to implement it:
- modified `clock_control_mcux_syscon.c` to provide frequency for the MCLK signal
- extended the `codec` API to handle input and output routing, extra sample rates and configuration for stream directions (playback, capture, duplex)

hakehuang committed:
boards: mimxrt595_evk,mimxrt685_evk: Instantiate WM8904

Add Flexcomm signal sharing configuration for the mimxrt595s/cm33 and
mimxrt685s/cm33 SoCs. Properly instantiate the I3C peripheral for
interfacing with the WM8904's control port. Modify pinctrl DTs of
mimxrt595_evk/mimxrt595s/cm33 and mimxrt685_evk/mimxrt685s/cm33.
Instantiate the WM8904 codec for said boards.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Yves Vandervennet <yves.vandervennet@nxp.com>
Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

hakehuang committed:
soc: mimxrt595s/cm33, mimxrt685s/cm33: Add clock config for audio

Add clock config for Flexcomm peripherals functioning as I2S interfaces.
Add MCLK clock config for the WM8904 codec located on the mimxrt595_evk
and mimxrt685_evk boards.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Yves Vandervennet <yves.vandervennet@nxp.com>
Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

hakehuang committed:
samples: i2s: Add i2s_codec example

This example demonstrates I2S audio playback combined with the codec
API. It can use the dmic API for audio input.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
drivers: audio: Add the wm8904 driver

Add driver for the Wolfson WM8904 audio codec.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Yves Vandervennet <yves.vandervennet@nxp.com>
Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
drivers: clock_control_mcux_syscon: Add I2S MCLK

Add support for the I2S MCLK signal (provided by CLOCK_GetMclkClkFreq).

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

marc-hb commented:
I added @dbaluta and @iuliana-prodan as reviewers. I didn't realize this went over the 15 reviewers limit. When this happens, Github normally warns about this limit and blocks the addition. But that time Github dropped random other reviewers instead! Temporary glitch? You have been warned.

I tried to cleanup that mess, hopefully the list is back to where it was.

hakehuang commented on samples/drivers/i2s/i2s_codec/sample.yaml:
```diff
@@ -0,0 +1,13 @@
+tests:
+  sample.drivers.i2s.codec:
+    tags: i2s
+    filter: dt_alias_exists("i2s_codec")
```
this filter will make the ci skip this case, and as you have add the platform_allow, either add this alias to dts or remove it.

VitekST replied:
Will remove then.

dbaluta commented on include/zephyr/audio/codec.h:
```diff
@@ -34,56 +34,66 @@ extern "C" {
+	AUDIO_PCM_RATE_96K = 96000,   /**< 96 kHz sample rate */
+	AUDIO_PCM_RATE_192K = 192000, /**< 192 kHz sample rate */
 } audio_pcm_rate_t;
 
```
I don't undertand this change? You seem to add some tabs / space. This seems not to be related with the subject of the patch.  api: codec: Extend API 

If you want to do some cleanups then it should be a separate patch. Otherwise, this doesn't bring any value as it is in this patch.

One patch should mostly do one thing!

VitekST replied:
I just invoked `clang-format` as recommended by contribution guidelines (https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format) to take care of code formatting. Is this something that shouldn't be done when changing existing files?

dbaluta replied:
> I just invoked `clang-format` as recommended by contribution guidelines (https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format) to take care of code formatting. Is this something that shouldn't be done when changing existing files?

yes, this shouldn't be taken into consideration for existing code. Only for new code added. If you still want to make the change you can make it as a separate PR. 

But maybe that is a task after you get this merged.

dbaluta commented on drivers/audio/Kconfig.wm8904:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024, NXP
+# SPDX-License-Identifier: Apache-2.0
```
Official way NXP adds copyright is. `Copyright 2024 NXP`.

dbaluta commented on soc/nxp/imxrt/imxrt5xx/cm33/soc.c:
```diff
@@ -85,6 +85,7 @@ const clock_frg_clk_config_t g_frg12Config_clock_init = {
 #endif
 
+
 /* System clock frequency. */
```
There is no need for a new line here.

VitekST committed:
api: codec: Extend API

Add sample rates. Add dai_route_t enum to allow codec configuration for
duplex operation. Allow for audio inputs to be controlled. Add functions
for signal routing.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

hakehuang commented:
CI board testing PASS.

hongshui3000 commented on include/zephyr/audio/codec.h:
```diff
@@ -35,7 +35,9 @@ extern "C" {
 	AUDIO_PCM_RATE_8K	= 8000,		/**< 8 kHz sample rate */
+	AUDIO_PCM_RATE_11K	= 11025,	/**< 11.025 kHz sample rate */
 	AUDIO_PCM_RATE_16K	= 16000,	/**< 16 kHz sample rate */
+	AUDIO_PCM_RATE_22K	= 22050,	/**< 24 kHz sample rate */
```
The comment here should be 22.050kHz instead of 24 kHz

decsny commented on samples/drivers/i2s/i2s_codec/Kconfig:
```diff
@@ -0,0 +1,29 @@
+	int "sample rate"
+	default 48000
+	help
+	  sample rate
```
these help texts are not sufficient, remember this will show in the menuconfig ui for users, it should have proper capitalization and grammar too

decsny commented on soc/nxp/imxrt/imxrt6xx/cm33/soc.c:
you can squash the soc commit for the 685 to the soc commit for the 595, same with the board commit to the other board commit, since they are small and doing the same thing, it makes the git history more consumable and less commit objects in the metadata

decsny commented on samples/drivers/i2s/i2s_codec/Kconfig:
```diff
@@ -0,0 +1,31 @@
+	int "Sample rate"
+	default 48000
+	help
+	  Sample frequency of the system.
```
please be more specific. what is the "system"?

VitekST replied:
Well, the microcontroller along with the codec. The sample frequency is the only one configured. The codec is configured for that rate and samples are processed at that rate. Can't think of a better designation, honestly.

decsny commented on samples/drivers/i2s/i2s_codec/Kconfig:
```diff
@@ -0,0 +1,31 @@
+	  filled by data from the DMIC peripheral and played back by the I2S
+	  output perihperal.
+
+config SAMPLE_FREQ
```
namespace as I2S_SAMPLE_FREQ?

VitekST replied:
I wouldn't do that, is this frequency is used also for the DMIC peripheral.

butok commented on samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.overlay:
```diff
@@ -0,0 +1,24 @@
+
+&i3c0 {
+	clk-divider = <20>;
+
```
blank line

VitekST replied:
@butok Intentional - I like to semantically separate blocks of code.

dbaluta commented on dts/bindings/audio/wolfson,wm8904.yaml:
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023, NXP
+# SPDX-License-Identifier: Apache-2.0
+
```
same here: Use `Copyright 2024 NXP`

dbaluta commented on drivers/audio/wm8904.h:
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2023 NXP Semiconductor INC.
+ * All rights reserved.
+ *
```
Same here. Fix the copyright.

dbaluta commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,760 @@
+/*
+ * Copyright (c) 2023 NXP Semiconductor INC.
+ * All rights reserved.
+ *
```
Official NXP copyright header is `Copyright 2024 NXP` all

dbaluta commented on samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.conf:
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2023, NXP
+#
```
Fix copyright.

dbaluta commented on samples/drivers/i2s/i2s_codec/src/main.c:
```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2024 NXP
+ *
```
Fix copyright here also. No `(c)` in the NXP recommended way of adding copyright. Just use `Copyright NXP 2024`

iuliana-prodan commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	const uint16_t regval = (volume | 0b001000000) & 0b011111111;
+	const uint16_t mask = 0b011111111;
+
+	switch (channel) {
```
This switch I would write it like:
```
uint8_t reg;
switch (channel) {
case AUDIO_CHANNEL_FRONT_LEFT:
	reg = WM8904_ANALOG_OUT2_LEFT;
	break;

case AUDIO_CHANNEL_FRONT_RIGHT:
	reg = WM8904_ANALOG_OUT2_RIGHT;
	break;

case AUDIO_CHANNEL_HEADPHONE_LEFT:
	reg = WM8904_ANALOG_OUT1_LEFT;
	break;

case AUDIO_CHANNEL_HEADPHONE_RIGHT:
	reg = WM8904_ANALOG_OUT1_RIGHT;
	break;

case AUDIO_CHANNEL_ALL:
	/* Set volume values with VU = 0 */
	reg = WM8904_ANALOG_OUT1_LEFT | WM8904_ANALOG_OUT1_RIGHT | WM8904_ANALOG_OUT2_LEFT | WM8904_ANALOG_OUT2_RIGHT;
	break;
default:
	return -EINVAL;
}

wm8904_update_reg(dev, reg, mask, regval);
return 0;
```

I see you do like this, here: https://github.com/zephyrproject-rtos/zephyr/pull/75943/commits/6f41b45b2320d72424dd58fe188eeb257da57228#diff-abe7424de1cefb3ba684964873bb5830e03f03ec4dd059ea2c67c07a8102548aR376

VitekST replied:
That can't be done, as the `reg` variable holds an ordinal register index, it isn't a flag-style variable with disjunctive bits.

iuliana-prodan replied:
ok, got it.

iuliana-prodan commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	 * [5:0] - VOL: leaving intact
+	 */
+
+	const uint16_t regval = ((mute << 8) | 0b001000000);
```
same as above

iuliana-prodan commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	 * [4:0] - VOL: 5 bit volume value
+	 */
+
+	const uint16_t regval = volume & 0b00011111;
```
same as above

iuliana-prodan commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	 * [4:0] - VOL: leaving intact
+	 */
+
+	const uint16_t regval = ((mute << 7) & 0b10000000);
```
same here

iuliana-prodan commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+}
+
+static void wm8904_start_output(const struct device *dev)
+{
```
Why are these 2 functions empty? 
No logging info, nothing?

VitekST replied:
@iuliana-prodan These functions correspond to the `audio_codec_start_output` and `audio_codec_stop_output` syscalls in `include/zephyr/audio/codec.h`. These syscall functions don't permit NULL values in the API structure, thus these functions are mandatory to implement. My goal was to change the `codec` API as little as possible, so that's why I left it as it is.

iuliana-prodan replied:
ok.

iuliana-prodan commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	 * Set VU = 1 for all output channels, VU takes effect for the whole
+	 * channel pair.
+	 */
+	wm8904_update_reg(dev, WM8904_ANALOG_OUT1_LEFT, 0b010000000, 0b010000000);
```
Use macros

iuliana-prodan commented on samples/drivers/i2s/i2s_codec/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_I2S=y
+#CONFIG_LOG=y
```
If you don't use it, delete it

VitekST commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	wm8904_update_reg(dev, WM8904_AUDIO_IF_1, (0x000CU), ((uint16_t)(word_size) << 2U));
+	{
+		uint16_t reg_val = 0;
+
```
It's not, it's just an artifact that was left unattended.

VitekST commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,759 @@
+	 * [5:0] - VOL: 6-bit volume value
+	 */
+
+	const uint16_t regval = (volume | 0b001000000) & 0b011111111;
```
@iuliana-prodan Do they? I personally don't see how that would be beneficial. The mask and content values differ practically for each register update operation, they are based directly on the codec's datasheet and, at least the ones I've added, are annotated. 

Does your comment apply also for other invocations of `wm8904_update_reg`, which here originally authored by Hake?

VitekST committed:
boards: mimxrt685_evk/mimxrt685s/cm33: Enable DMIC

Enable DMIC clock in soc.c - attach to chip's audio PLL. Add pinmux
definitions for the DMIC peripheral. Add nodes to SoC's device tree for
the DMIC peripheral and its audio channels. Configure the DMIC
peripheral in board's device tree to enable audio capture.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

dbaluta commented:
@VitekST  looks good to merge! We tested it on imxrt 685 and works fine.

VitekST commented:
@anangl Is everything up to your standards?

anangl commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,728 @@
+
+	const struct wm8904_driver_config *const dev_cfg = DEV_CFG(dev);
+
+	int error = pinctrl_apply_state(dev_cfg->pincfg, PINCTRL_STATE_DEFAULT);
+
+	if (error) {
+		LOG_ERR("pinctrl_apply_state fail: %d", error);
+		return error;
```
This makes the "default" pinctrl state required but "pinctrl-0" and "pinctrl-names" properties are not marked so in the binding. But is actually the external codec driver the one that should take care of configuring SoC pins? This looks a bit strange. @gmarull

VitekST replied:
@anangl Looks like another artifact from the previous author. I am also inclined to keep pinctrl configuration in the realm of drivers driving peripherals that expose used signals instead of the consumers and personally I don't see any reason not to configure the MCLK signal even though it wasn't done so before. Thus removed.

anangl commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,728 @@
+		return error;
+	}
+
+	error = i2c_recover_bus(dev_cfg->i2c.bus);
```
Why does this driver always need to recover the I2C bus?

VitekST replied:
@anangl Looks like a requirement when trying to use RT595/RT685's I3C peripheral as an I2C master.

anangl commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,728 @@
+	const uint16_t val = WM8904_REGVAL_OUT_VOL(mute, 0, 0, 0);
+	const uint16_t mask = WM8904_REGMASK_OUT_MUTE;
+
+	switch (channel) {
```
This `switch' is the same as in `wm8904_out_mute_config()`. Wouldn't it make sense to use a common part of code for both functions?

anangl commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,728 @@
+	const uint16_t val = WM8904_REGVAL_IN_VOL(mute, 0);
+	const uint16_t mask = WM8904_REGMASK_IN_MUTE;
+
+	switch (channel) {
```
Similarly here. It is the same as in `wm8904_in_mute_config()`.

VitekST replied:
@anangl Personally I think that the code can be followed a bit more easily than if one strives to avoid even two-fold repetitions at all cost, but if you insist...

anangl commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,728 @@
+	}
+}
+
+static void wm8904_update_reg(const struct device *dev, uint8_t reg, uint16_t mask, uint16_t val)
+{
+	uint16_t reg_val = 0;
+	uint16_t new_value = 0;
+
+	if (reg == 0x19) {
+		LOG_INF("try write mask %#x val %#x", mask, val);
```
These `LOG_INF()` calls in this function will probably generate a lot of logging messages. Wouldn't `LOG_DBG()` be a better choice?

anangl commented on boards/nxp/mimxrt595_evk/board.c:
```diff
@@ -279,34 +279,32 @@ static int mimxrt595_evk_init(void)
 
 #ifdef CONFIG_I2S
-
+#ifdef CONFIG_I2S_TEST_SEPARATE_DEVICES
```
This Kconfig option is only used by I2S tests. It's quite weird that it is handled in some board initialization code.

VitekST replied:
@anangl It's true that this property could be named differently (as I didn't author that name), I believe it has its place in either soc.c or board.c. This property implies how the signals between Flexcomm peripherals in an I2S mode should be shared. Duplex audio on the RTxxx devices with a single codec is handled by configuring one peripheral as an I2S master and the other as a slave. The slave peripheral then has the LRCK (WS) and SCK signals as inputs. It may be desirable not to set up signal sharing, for example when these devices are used with separate ADCs and DACs.

VitekST commented on drivers/audio/wm8904.c:
```diff
@@ -0,0 +1,728 @@
+	wm8904_in_mute_config(dev, AUDIO_CHANNEL_ALL, false);
+}
+
+static const struct audio_codec_api wm8904_driver_api = {.configure = wm8904_configure,
+							 .start_output = wm8904_start_output,
```
That is a product of `clang-format`. Am I not supposed to use it at all for new files, despite recommendations in contributor's guidelines?
